### ISSUE TITLE
Eureka server gives 404 on GET or DELETE operation with instanceID

### ISSUE THREAD

**Issue by dearjadu** [spring-cloud/spring-cloud-netflix#644]
I have a Spring Eureka Server running, which serves other Eureka clients as expected.  I am working on a set of Shell scripts that could (1) generate report of all available services, and (2) help clean up Eureka registry when in certain circumstances Eureka does not clean up the service references (happens in 5% of cases).  For this, I am using `cURL` command to make `GET` and `DELETE` operations.  `GET` operation works as expected as I am only querying by the `appID`.  However, any time I try to use the `instanceID`, it fails. This is true for `GET` and `DELETE` both, but I `GET` is not a big issue for me as I can always get all instances for an `appID` and iterate over them to find the target instance.  However, `DELETE` only supports invocations on `appID/instanceID` and hence it fails.  Here are the invocations using `cURL` that I am trying:

    $ curl http://eurekaserver:8761/eureka/apps/com-example-foo # com-example-foo is the `appID`
    WORKS - I get an XML output with app detail back

    $ curl http://eurekaserver:8761/eureka/apps/com-example-foo/1234 # 1234 is the `instanceID`
    FAILS - output as below:
    {
        "timestamp":1447621695429,
        "status":404,
        "error":"Not Found",
        "message":"Not Found",
        "path":"/eureka/apps/com-example-foo/111"
     }

    $ curl -X "DELETE" http://eurekaserver:8761/eureka/apps/com-example-foo/1234
    FAILS - same output as above

My Eureka Server is running as a Spring Boot (v1.3.0.M2) application using ``org.springframework.cloud:spring-cloud-starter-parent:Angel.SR3`` BOM.

**UPDATE**

FYI - the Eureka REST Operation reference is available at https://github.com/Netflix/eureka/wiki/Eureka-REST-operations

**Comment by spencergibb** [comment]
@dearjadu 
> appID is the name of the application and instanceID is the unique id associated with the instance. In AWS cloud, instanceID is the instance id of the instance and in other data centers, it is the hostname of the instance.

You're probably using the wrong instance id if you are getting a 404?

**Comment by dearjadu** [comment]
I'm using the instanceID that the GET reaponse gives me.

**Comment by dearjadu** [comment]
OK. So if it's not the instanceID that I supplied when I registered using the POST URL and if it is not the one that I 'GET' back when I ask Eureka for it, then what is it? Is there a way to get **that** instanceID?

**Comment by dearjadu** [comment]
Finally, I have figured out how the `cancel` operation can be invoked using REST URLs of a Eureka server. This works for Spring Cloud Eureka server, but should also work for the Netflix Eureka server.

The URL pattern for the `cancel` operation is as the following:

    DELETE http://eureka_host:eureka_port/eureka/apps/<appName>/<instanceId>

This is how it is documented at the [Eureka REST operations][1] page, but there's very little clarity as to what `<instanceId>` was supposed to be. As per the documentation, `<instanceId>` is the **hostname** of the host that runs the Eureka client. That did **not** work (IP address or host name). I tried passing in the same value that the `GET` URL gave me (for example, 192.168.55.55) or `localhost`.  That did not work either. I also tried passing in the `instanceId` value from the `GET` output (which would be the same as the value of eureka.instance.metadataMap.instanceId property). That too, didn't work.  I literally had to try different combinations to find this out. The `<instanceId>` is the concatenation of the hostname and instance ID, separated by `:`.  For example, `192.168.55.55:foo-app-some-random-str`.

Here's an example output of the `GET` operation listing the active instance registered with Eureka:

    <instance>
      <hostName>192.168.55.55</hostName>
      <app>FOO-APP</app>
      ...
      <metadata>
        <instanceId>foo-app-f4ea7b06fc03a05a06900713f7526a5d</instanceId>
      </metadata>
      ...
    </instance>

In this case, the `cancel` cURL command would look like this:

    $ curl -X "DELETE" http://eureka_host:eureka_port/eureka/apps/FOO-APP/192.168.55.55:foo-app-f4ea7b06fc03a05a06900713f7526a5d

That would de-register the instance as expected.

**That said,** I must confess that I wasn't paying much attention to the Eureka server logs. When you register the Eureka client, the log printed out the fully qualified name of the instance (`FOO-APP/192.168.55.55:foo-app-f4ea7b06fc03a05a06900713f7526a5d`), which I could have used as my guess.

I hope someone fixes this in the [Eureka documentation] [1].


  [1]: https://github.com/Netflix/eureka/wiki/Eureka-REST-operations
